core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -131,6 +135,11 @@ public void onComplete() {
                     }
                     // Releasing the lock to move ahead with the next request in queue.
                     releasePartitionLocks(shareFetchData.groupId(), topicPartitionData.keySet());
+                    // If we have a fetch request completed for a topic-partition, we release the locks for that partition,
+                    // then we should check if there is a pending share fetch request for the topic-partition and complete it.
+                    // We add the action to delayed actions queue to avoid an infinite call stack, which could happen if
+                    // we directly call delayedShareFetchPurgatory.checkAndComplete
+                    replicaManager.addCompleteDelayedShareFetchPurgatoryAction(CollectionConverters.asScala(result.keySet()).toSeq(), shareFetchData.groupId(), delayedShareFetchPurgatory);
```

Comment:
It's a bit weird to add the delayedAction through ReplicaManager. Perhaps we could create DelayedActionQueue in KafkaApis and pass it to both ReplicaManager and DelayedShareFetch.